Mathlib/Tactic/Module.lean

```diff
@@ -0,0 +1,641 @@
+up by the function `Mathlib.Tactic.Module.postprocess`. -/
+def matchScalarsAux (g : MVarId) : AtomM (List MVarId) := do
+  /- Parse the goal as an equality in a type `M` of two expressions `lhs` and `rhs`, with `M`
+  carrying an `AddCommMonoid` instance. -/
+  let eqData ← do
+    match (← g.getType').eq? with
+    | some e => pure e
+    | none => throwError "goal {← g.getType} is not an equality"
+  let .sort v₀ ← whnf (← inferType eqData.1) | unreachable!
+  let some v := v₀.dec | unreachable!
+  let ((M : Q(Type v)), (lhs : Q($M)), (rhs :Q($M))) := eqData
+  let iM ← synthInstanceQ q(AddCommMonoid.{v} $M)
+  /- Construct from the `lhs` expression a term `l₁` of type `qNF R₁ M` for some semiring `R₁` --
+  that is, a list of `(Q($R₁) × Q($M)) × ℕ`s (two `Expr`s and a natural number) -- together with a
+  proof that `lhs` is equal to the `R₁`-linear combination in `M` this represents. -/
+  let e₁ ← parse iM lhs
+  have u₁ : Level := e₁.fst
+  have R₁ : Q(Type u₁) := e₁.snd.fst
+  have _iR₁ : Q(Semiring.{u₁} $R₁) := e₁.snd.snd.fst
+  let iRM₁ ← synthInstanceQ q(Module $R₁ $M)
+  assumeInstancesCommute
+  have l₁ : qNF R₁ M := e₁.snd.snd.snd.snd.fst
+  let pf₁ : Q($lhs = NF.eval $(l₁.toNF)) := e₁.snd.snd.snd.snd.snd
+  /- Do the same for the `rhs` expression, obtaining a term `l₂` of type `qNF R₂ M` for some
+  semiring `R₂`. -/
+  let e₂ ← parse iM rhs
+  have u₂ : Level := e₂.fst
+  have R₂ : Q(Type u₂) := e₂.snd.fst
+  have _iR₂ : Q(Semiring.{u₂} $R₂) := e₂.snd.snd.fst
+  let iRM₂ ← synthInstanceQ q(Module $R₂ $M)
+  have l₂ : qNF R₂ M := e₂.snd.snd.snd.snd.fst
+  let pf₂ : Q($rhs = NF.eval $(l₂.toNF)) := e₂.snd.snd.snd.snd.snd
+  /- Lift everything to the same scalar ring, `R`. -/
+  let ⟨_, _, _, iRM, ⟨l₁', pf₁'⟩, ⟨l₂', pf₂'⟩, _⟩ ← qNF.matchRings iRM₁ _ iRM₂ l₁ l₂ q(0) q(0)
+  /- Construct a list of goals for the coefficientwise equality of these formal linear combinations,
+  and resolve our original goal (modulo these new goals). -/
+  let (mvars, pf) ← reduceCoefficientwise iRM l₁' l₂'
+  g.assign q(NF.eq_of_eval_eq_eval $pf₁ $pf₂ $pf₁' $pf₂' $pf)
+  return mvars
+
+/-- Lemmas used to post-process the result of the `match_scalars` and `module` tactics by converting
+the `algebraMap` operations which (which proliferate in the constructed scalar goals) to more
+familiar forms: `ℕ`, `ℤ` and `ℚ` casts. -/
+def algebraMapThms : Array Name := #[``eq_natCast, ``eq_intCast, ``eq_ratCast]
+
+/-- Postprocessing for the scalar goals constructed in the `match_scalars` and `module` tactics.
+These goals feature a proliferation of `algebraMap` operations (because the scalars start in `ℕ` and
+get successively bumped up by `algebraMap`s as new semirings are encountered), so we reinterpret the
+most commonly occuring `algebraMap`s (those out of `ℕ`, `ℤ` and `ℚ`) into their standard forms (`ℕ`,
+`ℤ` and `ℚ` casts) and then try to disperse the casts using the various `push_cast` lemmas. -/
+def postprocess (mvarId : MVarId) : MetaM MVarId := do
```

Comment:
Could post processing also use `mul_one`, to clean up the awkward looking `x * 1` that `match_scalars` produces? (I appreciate that `ring` doesn't care.)

Reply:
Still thinking about this one -- my mild hesitation is that it would also mean that any `* 1`s in the original also get cleared, e.g. `match_scalars (2 * 1 * 1) • x = a • x` produces a goal `2 = a` rather than `2 * 1 * 1 = a`.  This is probably not a big deal, but I'm going to test how painful it is to carry around an `Option Q($R)` rather than a `Q($R)` in order to distinguish between these two cases.